### PR TITLE
[FlashInfer v0.6.6][RL] Support fp8-last-n-bf16 RL for `flashinfer_trtllm_routed` moe backend

### PR DESCRIPTION
**Pull request by zianglih**
## Motivation
@humansand

This PR is the last missing piece of Miles Blackwell mxfp8 RL training: https://github.com/radixark/miles/issues/615 , https://github.com/radixark/miles/pull/614 .

Dependencies:
- FlashInfer `trtllm_bf16_routed_moe`: https://github.com/flashinfer-ai/flashinfer/pull/2594
- Earlier FlashInfer routed moe integration: https://github.com/sgl-project/sglang/pull/19537
- Mixed mxfp8 + bf16 serving: https://github.com/sgl-project/sglang/pull/18742

FlashInfer v0.6.5 has a bug https://github.com/flashinfer-ai/flashinfer/pull/2697#issuecomment-4028767021, so this PR cannot be merged until v0.6.6 release.

## Modifications

- Integrate bf16 `trtllm_bf16_routed_moe` from FlashInfer v0.6.6, mirroring existing fp8/mxfp8 code path in https://github.com/sgl-project/sglang/pull/19537
- Expand test coverage in `test/registered/backends/test_flashinfer_trtllm_gen_moe_backend.py`
- Add a weight update test for mxfp8-last-n-bf16 checkpoint + flashinfer_trtllm_routed moe backend in `test/registered/rl/test_update_weights_blackwell.py`

## Accuracy Tests

```
root@B200-78:/sgl-workspace/sglang# python3 -m pytest -q test/registered/backends/test_flashinfer_trtllm_gen_moe_backend.py
......                                                                                                         [100%]
================================================== warnings summary ==================================================
../../usr/local/lib/python3.12/dist-packages/_pytest/config/__init__.py:1428
  /usr/local/lib/python3.12/dist-packages/_pytest/config/__init__.py:1428: PytestConfigWarning: Unknown config option: asyncio_mode
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

registered/backends/test_flashinfer_trtllm_gen_moe_backend.py::TestFlashinferTrtllmGenMoeBackendFP8::test_gsm8k
registered/backends/test_flashinfer_trtllm_gen_moe_backend.py::TestFlashinferTrtllmGenMoeBackendMXFP8::test_gsm8k
registered/backends/test_flashinfer_trtllm_gen_moe_backend.py::TestFlashinferTrtllmGenMoeBackendBF16::test_gsm8k
registered/backends/test_flashinfer_trtllm_gen_moe_backend.py::TestFlashinferTrtllmGenMoeBackendFP8Routed::test_gsm8k
registered/backends/test_flashinfer_trtllm_gen_moe_backend.py::TestFlashinferTrtllmGenMoeBackendMXFP8Routed::test_gsm8k
registered/backends/test_flashinfer_trtllm_gen_moe_backend.py::TestFlashinferTrtllmGenMoeBackendBF16Routed::test_gsm8k
  /sgl-workspace/sglang/python/sglang/test/few_shot_gsm8k.py:54: DeprecationWarning: Including the scheme in --host ('http://127.0.0.1') is deprecated. Pass just the hostname (e.g. '127.0.0.1') instead.
    set_default_backend(RuntimeEndpoint(normalize_base_url(args.host, args.port)))

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
6 passed, 7 warnings in 615.78s (0:10:15)
```

```
root@B200-78:/sgl-workspace/sglang# python3 -m pytest -q test/registered/rl/test_update_weights_blackwell.py
...                                                                                                   [100%]
============================= warnings summary =============================
../../usr/local/lib/python3.12/dist-packages/_pytest/config/__init__.py:1428
  /usr/local/lib/python3.12/dist-packages/_pytest/config/__init__.py:1428: PytestConfigWarning: Unknown config option: asyncio_mode
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
3 passed, 1 warning, 9 subtests passed in 442.51s (0:07:22)
```
## Benchmarking and Profiling

For softmax MoE models like `Qwen3-30B-A3B`, enabling piecewise CUDA graph causes instable numerics, which might be relevant to torch compiling: https://github.com/sgl-project/sglang/blob/538acb4c46e002738b9abd3e56571b9b773cc627/python/sglang/srt/layers/moe/topk.py#L447-L463

bf16 checkpoint with `flashinfer_trtllm` (baseline)
```
python -m sglang.launch_server --kv-cache-dtype bf16 --model Qwen/Qwen3-30B-A3B-Instruct-2507 --fp8-gemm-backend flashinfer_trtllm --moe-runner-backend flashinfer_trtllm &
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1209 --parallel 1209 --platinum
Accuracy: 0.967
Invalid: 0.000
Latency: 8.572 s
Output throughput: 19904.440 token/s
Accuracy: 0.967
Invalid: 0.000
Latency: 8.504 s
Output throughput: 20070.594 token/s
Accuracy: 0.967
Invalid: 0.000
Latency: 8.234 s
Output throughput: 20726.172 token/s
```

bf16 checkpoint with `flashinfer_trtllm_routed`
```
python -m sglang.launch_server --kv-cache-dtype bf16 --model Qwen/Qwen3-30B-A3B-Instruct-2507 --fp8-gemm-backend flashinfer_trtllm --moe-runner-backend flashinfer_trtllm_routed &
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1209 --parallel 1209 --platinum
Accuracy: 0.965
Invalid: 0.000
Latency: 9.317 s
Output throughput: 18291.868 token/s
Accuracy: 0.966
Invalid: 0.000
Latency: 9.254 s
Output throughput: 18426.035 token/s
Accuracy: 0.965
Invalid: 0.000
Latency: 8.928 s
Output throughput: 19086.286 token/s
```

bf16 checkpoint with `flashinfer_trtllm_routed` with `--disable-piecewise-cuda-graph`
```
python -m sglang.launch_server --kv-cache-dtype bf16 --model Qwen/Qwen3-30B-A3B-Instruct-2507 --fp8-gemm-backend flashinfer_trtllm --moe-runner-backend flashinfer_trtllm_routed --disable-piecewise-cuda-graph &
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1209 --parallel 1209 --platinum
Accuracy: 0.965
Invalid: 0.000
Latency: 9.526 s
Output throughput: 17901.575 token/s
Accuracy: 0.965
Invalid: 0.000
Latency: 9.278 s
Output throughput: 18392.800 token/s
Accuracy: 0.965
Invalid: 0.000
Latency: 8.978 s
Output throughput: 18991.882 token/s
```

mxfp8-last-n-bf16 checkpoint
```
python -m sglang.launch_server --kv-cache-dtype bf16 --model zianglih/Qwen3-30B-A3B-Instruct-2507-MXFP8-last-8-BF16 --fp8-gemm-backend flashinfer_trtllm --moe-runner-backend flashinfer_trtllm_routed &
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1209 --parallel 1209 --platinum
Accuracy: 0.970
Invalid: 0.000
Latency: 10.267 s
Output throughput: 16594.531 token/s
Accuracy: 0.971
Invalid: 0.000
Latency: 10.003 s
Output throughput: 17022.217 token/s
Accuracy: 0.969
Invalid: 0.000
Latency: 9.421 s
Output throughput: 18088.243 token/s
```
mxfp8-last-n-bf16 checkpoint with `--disable-piecewise-cuda-graph`
```bash
python -m sglang.launch_server --kv-cache-dtype bf16 --model zianglih/Qwen3-30B-A3B-Instruct-2507-MXFP8-last-8-BF16 --fp8-gemm-backend flashinfer_trtllm --moe-runner-backend flashinfer_trtllm_routed --disable-piecewise-cuda-graph &
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1209 --parallel 1209 --platinum
Accuracy: 0.970
Invalid: 0.000
Latency: 9.863 s
Output throughput: 17271.128 token/s
Accuracy: 0.970
Invalid: 0.000
Latency: 9.806 s
Output throughput: 17368.509 token/s
Accuracy: 0.970
Invalid: 0.000
Latency: 9.632 s
Output throughput: 17686.235 token/s
```
## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: test/registered/rl/test_update_weights_from_disk_mxfp8.py
Comment (zianglih) [inline test/registered/rl/test_update_weights_from_disk_mxfp8.py:1]: We want a dedicated blackwell test file because mxfp/nvfp data types usually requires an extra weight swizzling step after weight update. We want a separate file for testing these behaviors.

File: test/registered/rl/test_update_weights_blackwell.py [hunk 1]
```diff
@@ -0,0 +1,451 @@
+from sglang.test.ci.ci_register import register_cuda_ci
+
+register_cuda_ci(est_time=500, suite="stage-b-test-4-gpu-b200")
```
Comment (Fridge003) [inline test/registered/rl/test_update_weights_blackwell.py]: Maybe move it to `stage-c-test`. `stage-b-test` needs to be lightweight

File: test/registered/rl/test_update_weights_from_disk_mxfp8.py [hunk 2]
```diff
@@ -0,0 +1,451 @@
+from sglang.test.ci.ci_register import register_cuda_ci
+
+register_cuda_ci(est_time=500, suite="stage-b-test-4-gpu-b200")
+
```
Comment (Fridge003) [inline test/registered/rl/test_update_weights_from_disk_mxfp8.py:4]: Can we prune this test to maybe 200 seconds? 500 second is a little long

File: python/sglang/srt/layers/quantization/unquant.py [hunk 3]
```diff
@@ -319,12 +319,58 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+        param: torch.nn.Parameter,
+        weight_name: str,
+    ) -> None:
+        """Restore canonical BF16 MoE load shapes before hot weight copy.
+
+        The flashinfer TRT-LLM BF16 postprocess reshapes expert weights into
+        block layout. During weight update, checkpoint tensors are in
+        canonical layout and need a temporary shape restore for copy.
+        """
+        if not self.use_flashinfer_trtllm_moe:
```
Comment (Fridge003) [inline python/sglang/srt/layers/quantization/unquant.py]:
```suggestion
        if not get_moe_runner_backend().is_flashinfer_trtllm_routed():
```

File: test/registered/rl/test_update_weights_blackwell.py (hunk 1 above)
Comment (zianglih) [reply test/registered/rl/test_update_weights_blackwell.py]: Done by https://github.com/sgl-project/sglang/pull/20214/commits/a5ae1b3616a29ca6ae5df51ef4b704838912e2df

File: test/registered/rl/test_update_weights_from_disk_mxfp8.py (hunk 2 above)
Comment (zianglih) [reply test/registered/rl/test_update_weights_from_disk_mxfp8.py:4]: Done by https://github.com/sgl-project/sglang/pull/20214/commits/a5ae1b3616a29ca6ae5df51ef4b704838912e2df

**Review by Fridge003 — APPROVED** [review]
